### PR TITLE
feat: updated to Next 13 and synced with tRPC repo

### PR DESCRIPTION
**Pull request by PiotrekPKP**
This pull request:
- updates Next version to 13.0.1
- updates tRPC versions to the latest RC
- adds a bunch of helpers for tRPC routers and procedures, as well as inference helpers for front-end code (basically just copied the code tRPC examples and ct3a had :+1: )

### CODE REVIEW

File: apps/nextjs/package.json [hunk 1]
```diff
@@ -12,26 +12,26 @@
...
     "@acme/db": "*",
-    "@tanstack/react-query": "^4.7.2",
-    "@trpc/client": "10.0.0-proxy-beta.13",
-    "@trpc/next": "10.0.0-proxy-beta.13",
-    "@trpc/react": "10.0.0-proxy-beta.13",
-    "@trpc/server": "10.0.0-proxy-beta.13",
-    "next": "12.3.1",
-    "react": "18.1.0",
-    "react-dom": "18.1.0",
+    "@tanstack/react-query": "^4.3.8",
```
Comment (juliusmarminge) [inline apps/nextjs/package.json:15]: Why down rather than up?

File: apps/nextjs/package.json [hunk 2]
```diff
@@ -12,26 +12,26 @@
...
+    "@types/react": "^18.0.24",
+    "@types/react-dom": "^18.0.8",
     "@typescript-eslint/eslint-plugin": "^5.33.0",
     "@typescript-eslint/parser": "^5.33.0",
     "autoprefixer": "^10.4.7",
     "eslint": "8.22.0",
-    "eslint-config-next": "12.3.1",
-    "next-transpile-modules": "^9.0.0",
+    "eslint-config-next": "13.0.1",
+    "next-transpile-modules": "^10.0.0",
```
Comment (juliusmarminge) [inline apps/nextjs/package.json:34]: Next has (experimental), built in transpiling in 13. Should we try it? 

https://github.com/vercel/next.js/blob/3cb0c297c89fde815a34f23528ea41d211d57e3c/packages/next/server/config-shared.ts#L163

File: packages/api/src/trpc.ts [hunk 3]
```diff
@@ -8,3 +8,7 @@ export const t = initTRPC.context<Context>().create({
     return shape;
   },
 });
+
```
Comment (juliusmarminge) [inline packages/api/src/trpc.ts:11]: Dont export root `t` object

**Review by juliusmarminge — COMMENTED** [review]
Thanks 🙏. Left some thought

**Comment by PiotrekPKP** [comment]
@juliusmarminge thanks for the review! I'll see what I can do with Next's transpile modules alternative. Also, I was having troubles with getting the query package working and forgot to bump the versions, sorry! I'll fix those issues 😉

**Comment by PiotrekPKP** [comment]
@juliusmarminge Hey, I've updated the pull request with proposed (and new) changes:
- fixed the `@tanstack/react-query` version to be the latest
- added the same logic as in `ct3a` for the env validation (that required `es2017` so that's why it's bumped up in `tsconfig.json`)
- replaced `next-transpile-modules` with the new experimental config in `next.config.mjs` - everything including hot-reload worked for me after the change, so I suppose it's ok not to use the external package anymore
- fixed the unnecessary `t` root object export

File: apps/expo/package.json [hunk 4]
```diff
@@ -13,16 +13,16 @@
   "dependencies": {
     "@acme/api": "*",
     "@shopify/flash-list": "^1.2.1",
-    "@tanstack/react-query": "^4.7.2",
-    "@trpc/client": "10.0.0-proxy-beta.13",
-    "@trpc/react": "10.0.0-proxy-beta.13",
-    "@trpc/server": "10.0.0-proxy-beta.13",
+    "@tanstack/react-query": "^4.14.1",
+    "@trpc/client": "10.0.0-rc.3",
+    "@trpc/react-query": "10.0.0-rc.3",
```
Comment (filipweilid) [inline apps/expo/package.json:18]: Hey, noticed that the trpc util for expo hasn't been updated to match this change
https://github.com/t3-oss/create-t3-turbo/blob/f30c8af893f11f0145c92b4845dd9bb14cfe4df4/apps/expo/src/utils/trpc.tsx#L1

**Review by juliusmarminge — APPROVED** [review]
As Filip pointed out the expo utils aren't updated. Fix that and I'll try and test this out tomorrow and merge 🚀

**Comment by PiotrekPKP** [comment]
Fixed, sorry for the mistake!

**Comment by juliusmarminge** [comment]
Hey @PiotrekPKP I just tested this locally and seems to mostly work fine.

I do however get an error regarding the React version:

CleanShot 2022-11-05 at 11 10 06

There's a small dilemma here;
- Next.js 13 requires 18.2
- React Native requires 18.1

From experience, mixing React versions can be a bit of a hit & miss and often causes errors due to context/rules-of-hooks etc.

What do you think we should do?

**Comment by juliusmarminge** [comment]
Just noticed there's a release candidate for React Native with 18.2 support, tested it and seems to work fine. Should we go for it?

**Comment by PiotrekPKP** [comment]
Strange, I did not get a single time hitting that error. Anyway I think that rc is the best option right now @juliusmarminge

**Comment by juliusmarminge** [comment]
> Strange, I did not get a single time hitting that error. Anyway I think that rc is the best option right now @juliusmarminge

To be clear, I didn't get any errors now either, but from experience they are deemed to come if you keep React versions mismatched.

I didn't have `manypkg` setup in CI so that's why it wasn't caught. Set it up now hence it's failing now:

CleanShot 2022-11-05 at 12 05 35

**Comment by PiotrekPKP** [comment]
I'm updating to the new rc for RN right now, will try to push it soon if everything will work on my side

**Comment by PiotrekPKP** [comment]
Run `pnpm manypkg fix` and updated more packages on the RN side, SHOULD be fine now haha

**Comment by juliusmarminge** [comment]
Seems to work wonderful. Thank you for hanging by 🙏🏼

**Comment by PiotrekPKP** [comment]
> Seems to work wonderful. Thank you for hanging by 🙏🏼

No problem!
Thanks for this project, been using it for few weeks and works great!

**Pull request merged**